test/unittests/test_dd_synthesis_dc.cpp

```diff
@@ -62,7 +62,7 @@ TEST_P(TestDDSynthDc, GenericDDSynthesisDcTest) {
 
     // generate the complete truth table.
     extend(tt);
-    augmentWithConstants(tt, totalNoBits, true);
+    augmentWithConstants(tt, totalNoBits, false, true);
```

Comment:
yes. but you know (or should know) that in the first example, the first two input variables are constants and the last two output variables are garbage. In the same sense, the last two input variables in the second example are constant, while the last two output variables are garbage again.
All three TT have 2 primary inputs and 2 primary outputs and you should be able to compare two truth tables based on that. This is essentially what the `equals` function is for. Any pairs of these three truth tables should compare equal.

Reply:
our `equal` function requires the length of the inputs and outputs of both the truth tables to be the same. So we have to construct the truth table such that it can be compared right?

I'm getting confused here. Am I missing something trivial? Are you suggesting to alter the original irreversible function without `augmentWithConstants` function so that it can be compared with the circuit generated from the `qc`?

Reply:
I am suggesting to adapt the `equal` function so that it can handle the above scenario properly.
The `equal` function should either just compare for "total" equality (it already does this properly) or it should be capable of comparing two TTs on their primary inputs and outputs. The only requirement here is that two TTs need to have the same number of primary inputs and outputs in order to be comparable.

Reply:
In the very end, no `extend` or `augmentWithConstants` call should be necessary at all in the tests and it should only be necessary to do `EXPECT_TRUE(TruthTable::equal(tt, ttqc));`

Reply:
I think I still need to use `extend` it. The pla files from  the revlib library are in some wired format. 

for eg:

001-1          001
00111         010

from the above, you can see that 00111 can basically have 2 outputs.

Reply:
Two scenarios here: that PLA file is wrong, i.e. there should never be two input patterns that could be the same. Or, in that case the more specific input (the one without the DC) takes precedence.
To be honest, I am not too familiar with the pla format, but that should be something that can be determined.

Reply:
I'm pretty sure it should be the later one. I will be using `extend` then.

Reply:
Is the equals function enough?



`    auto TruthTable::equal(TruthTable& tt1, TruthTable& tt2, std::size_t const& noOfPrimaryInputs, std::size_t const& noOfPrimaryOutputs, bool append, bool equalityUpToDontCare) -> bool {
        if (!equalityUpToDontCare) {
            return (tt1 == tt2);
        }

        const auto nBits = std::max(tt1.nInputs(), tt2.nInputs());
        assert(nBits < 65U);
        const auto totalInputs = 1U << noOfPrimaryInputs;
        const auto ancillaBits = nBits - noOfPrimaryInputs;
        const auto matchingAncillaBits = noOfPrimaryOutputs - noOfPrimaryInputs;


        std::uint64_t n = 0U;

        while (n < totalInputs) {
            const auto input = TruthTable::Cube::fromInteger(n, noOfPrimaryInputs);
            ++n;

            TruthTable::Cube tt1Input{input};
            TruthTable::Cube tt2Input{input};

            if(tt1.nOutputs() > tt1.nInputs()){
                for(auto i = 0U; i < matchingAncillaBits ; i++){
                    tt1Input.insertZero();
                }
            }

            if(tt2.nOutputs() > tt2.nInputs()){
                for(auto i = 0U; i < matchingAncillaBits ; i++){
                    tt2Input.insertZero();
                }
            }

            const bool alterInput1 = tt2.nInputs() == noOfPrimaryInputs && tt1.nInputs() > noOfPrimaryInputs;
            const bool alterInput2 = tt1.nInputs() == noOfPrimaryInputs && tt2.nInputs() > noOfPrimaryInputs;

            for(auto i = 0U; i < ancillaBits ; i++){
                if(append){
                    if(alterInput1 && !alterInput2){
                        tt1Input.emplace_back(false);
                    }
                    else if(alterInput2 && !alterInput1){
                        tt2Input.emplace_back(false);
                    }
                }
                else{
                    if(alterInput1 && !alterInput2){
                        tt1Input.insertZero();
                    }
                    else if(alterInput2 && !alterInput1){
                        tt2Input.insertZero();
                    }
                }
            }


            const auto foundtt1 = tt1.find(tt1Input) != tt1.end();
            const auto foundtt2 = tt2.find(tt2Input) != tt2.end();

            if (!foundtt1 || !foundtt2) {
                continue;
            }

            TruthTable::Cube tt1Output{tt1[tt1Input]};
            tt1Output.resize(nBits);
            TruthTable::Cube tt2Output{tt2[tt2Input]};
            tt2Output.resize(nBits);

            if (!Cube::checkCubeEquality(tt1Output, tt2Output)) {
                return false;
            }
        }

        return true;
    }`

Reply:
To be honest, I don't really have time to check that now. The signature of the `equals` function should only require two `const&` TTs and a boolean flag whether to check for complete equivalence though.
The number of primary inputs and outputs should be something that is being stored in the TT data structure and be accessible from there.

In addition to the above, I am not 100% sure that the PLA to TT parser works as intended. Judging from the following example https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiJtvj_ler7AhUsSfEDHVjiDakQFnoECA4QAQ&url=http%3A%2F%2Fopencircuitdesign.com%2Fmagic%2Farchive%2Fpapers%2Fmpla.pdf&usg=AOvVaw2hce4Ie6URnkeDAfbW1EKd it does not seem that straightforward to just take the midterms and put them in the TT. Maybe cross-check that with revisit/cirkit.

Reply:
I also need a another boolean for `equal` right? The bit positions of the constants that are added to the encoding and the one-pass based synthesis technique are different. I don't think it is possible to check without including another boolean. 

I have made the necessary changes and made sure that the `equal` function is symmetric.

To be honest, I would prefer working on integrating the encoding without any additional line functionality.

Reply:
The TruthTable class should store which of its bits are constant/garbage. There might be way more general situations where constants are neither only at the end or only the beginning.
This will require some modifications in the TT class itself. However, these changes are overdue anyway.

You can certainly work on anything else in parallel.

Reply:
I want to confirm a few things. 

Just consider the irreversible truth table defined above.

I can determine the minimum number of ancilla/garbage bits to be added using the truth table. 

In fact, I can calculate the total number of bits as well.  

Now my question: without knowing the synthesis technique that is going to be used, how do I distinguish between the constant bit positions and primary input bit positions?

Reply:
When adding constants to a truth table, something in the truth table structure should be updated to reflect that a constant was added at a certain position. (Could be a `std::vector<bool> constants`)
In a similar fashion, garbage outputs should be "marked" when adding them.
For the irreversible TT, there are no constants and no garbage. For the synthesized TTs, the position of the constants and garbage depends on the technique used.
The equals method just filters out the constants and/or garbage outputs and compares the truth tables based on their primary inputs and outputs.

Reply:
You mean to say the `ttqc` obtained from the `buildCircuitToTruthtable` should be updated with the vectors of constants and garbages? 

Because currently the synthesize function (static) takes the const reference of the Truth table. Although, I can reflect the corresponding constants and garbage positions, it will not be reflected on the original truth table that is passed as an argument to the static synthesize function.

Reply:
The original truth table does not need to be changed in any way. 

The `QuantumComputation` class provides an `ancillary` and a `garbage` member that keep track of whether a Qubit is constant or garbage. These are set with corresponding class methods (something like `setLogicalQubitGarbage`). The synthesis methods should set these variables on the resulting quantum circuit accordingly.

Based on that the buildTT function that takes a circuit and constructs the TT should respect these constants and garbage values, i.e., any constant input should be only 0 during the simulation.
Probably, the buildTT should also store these arrays in the truth table class (so that they are available later).

The equals function should make sure that only primary outputs are compared (ignoring garbage outputs).

I hope this makes things clearer. In the end, it should be possible to just compare the original TT and the TT from the circuit without needing to specify any additional information and without extending any of the truth tables.

Reply:
I have made the necessary changes.

There is only one thing:

Irrespective of anything, we have to extend the truth table. Consider the pla file that you considered as a reference. 
We need to extend the `tt` to correctly calculate the corresponding I/O combination.  Without extending, we are not even comparing the right `tt`.

I hope this is makes sense.

Reply:
Thanks! I'll look over the changes over the course of the weekend.

I agree, that the TT has to be extended. However, I am wondering whether that should actually happen at the PLA import. Technically, the "TT" that is imported isn't really a truth table (but rather a collection of minterms). If the extension happens in the PLA import function, it should not be needed elsewhere.

Reply:
I agree. I will just move the `extent(tt)` function to the `readPla` function. Therefore, there will be no need to use the `extend(tt)` any where else. I hope this sorts out the issue.